src/webgpu/shader/validation/expression/call/builtin/smoothstep.spec.ts

```diff
@@ -0,0 +1,239 @@
+const builtin = 'smoothstep';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import {
+  Type,
+  concreteTypeOf,
+  elementTypeOf,
+  isFloatType,
+  kAllScalarsAndVectors,
+  kConvertableToFloatScalarsAndVectors,
+  scalarTypeOf,
+} from '../../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValuesTypes = objectsToRecord(kConvertableToFloatScalarsAndVectors);
+const kArgumentTypes = objectsToRecord(kAllScalarsAndVectors);
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() rejects invalid values
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValuesTypes))
+      .filter(u => stageSupportsType(u.stage, kValuesTypes[u.type]))
+      .beginSubcases()
+      .expand('value1', u => [-1000, -10, 0, 10, 1000])
+      .expand('value2', u => [-1000, -10, 0, 10, 1000])
+  )
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kValuesTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const type = kValuesTypes[t.params.type];
+
+    // We expect to fail if low == high as it results in a DBZ
+    const expectedResult = t.params.value1 !== t.params.value2;
+
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      expectedResult,
+      [type.create(t.params.value1), type.create(t.params.value2), type.create(0)],
+      t.params.stage,
+      /* returnType */ concreteTypeOf(type, [Type.abstractFloat, Type.f32, Type.f16])
+    );
+  });
+
+g.test('argument_types')
+  .desc(
+    `
+Validates that scalar and vector arguments are rejected by ${builtin}() if not float type or vecN<float type>
+`
+  )
+  .params(u => u.combine('type', keysOf(kArgumentTypes)))
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kArgumentTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const type = kArgumentTypes[t.params.type];
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      /* expectedResult */ isFloatType(elementTypeOf(type)),
+      [type.create(0), type.create(1), type.create(2)],
+      'constant',
+      /* returnType */ concreteTypeOf(type)
```

Comment:
```suggestion
      /* returnType */ concreteTypeOf(type, [Type.f32])
```

Reply:
Interesting, doing this actually exposed a bug in the test: `expectedResult` was true only when `isFloatType(elementTypeOf(type))`, but this was incorrect for `abstract-int` (and composites). This became obvious when attempting to assign the result to an `f32`, as the cases would be marked as failing, even thought they should pass. They were failing before that because we were attempting to assign the result to an `i32` (or composite). I added a new function `isConvertibleToFloatType`.

Reply:
Done.